File: core/WordSearch/WordSearch.java
package com.dbc;

public class WordSearch {
    private boolean dfs(int row, int col, int point, char[][] board, String word) {
        if (point == word.length() - 1 && word.charAt(point) == board[row][col]) return true;
        if (word.charAt(point) != board[row][col]) return false;
        board[row][col] = '*';
        int[][] dir = new int[][]{{row, col + 1}, {row, col - 1}, {row + 1, col}, {row - 1, col}};
        for (int[] newDir : dir) {
            if (newDir[0] >= 0 && newDir[0] < board.length && newDir[1] >= 0 && newDir[1] < board[0].length && dfs(newDir[0], newDir[1], point + 1, board, word)) return true;
        }
        board[row][col] = word.charAt(point);
        return false;
    }

    public boolean exist(char[][] board, String word) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (board[i][j] == word.charAt(0) && dfs(i, j, 0, board, word)) return true;
            }
        }
        return false;
    }
}

File: core/WordSearch/WordSearch.py
from typing import List


def exist(board: List[List[str]], word: str) -> bool:
    m, n, length = len(board), len(board[0]), len(word)

    def dfs(row, col, point):
        if point == length - 1 and board[row][col] == word[point]:
            return True
        if board[row][col] != word[point]:
            return False
        board[row][col] = "_"
        for new_row, new_col in [[row, col + 1], [row, col - 1], [row + 1, col], [row - 1, col]]:
            if 0 <= new_row < m and 0 <= new_col < n and dfs(new_row, new_col, point + 1):
                return True
        board[row][col] = word[point]
        return False

    for i in range(m):
        for j in range(n):
            if board[i][j] == word[0] and dfs(i, j, 0):
                return True

    return False


if __name__ == '__main__':
    print(exist([["a"]], "a"))

